backoffice/views.py

```diff
@@ -256,40 +257,69 @@ def user_list(request):
+        messages.success(request, _(u"Successfully reenabled {}'s account").format(user.username))
+    user_account.changed_by = request.user
+    user_account.standing_last_changed_at = timezone.now()
+    user_account.save()
+
+def change_password(request, user):
+    user.set_password(request.POST['new-password'])
+    user.save()
+    messages.success(request, _(u"User password changed"))
+
+def change_grade(request, user):
+    """ Change a certificate grade per user, per course.
+
+    Grade comes as string and needs to range from 0 to 1.
+    Args:
+         request (HttpRequest): Contains the course id in request.POST['course-id'].
+         user (User): The user attached to the certificate.
+     """
+    course = get_course_key(request.POST['course-id'])
+    generated_certificate = GeneratedCertificate.objects.get(user=user, course_id=course)
+    try:
+        if not 0 <= float(request.POST['new-grade']) <= 1:
+            messages.error(request, _(u"Grades range from 0 to 1."))
+        else:
+            generated_certificate.grade = request.POST['new-grade']
+            generated_certificate.save()
+            messages.success(request, _(u"User grade changed."))
+    except ValueError:
+        messages.error(request, _(u"Invalid certificate grade."))
+
+user_actions = {'ban-user' : ban_user,
+                'change-password' : change_password,
+                'change-grade' : change_grade}
+
 @group_required('fun_backoffice')
 def user_detail(request, username):
     try:
         user = User.objects.select_related('profile').get(username=username)
     except User.DoesNotExist:
         raise Http404()
-    if 'action' in request.POST:
-        if request.POST['action'] == 'ban-user':
-            user_account, _created = UserStanding.objects.get_or_create(
-                    user=user, defaults={'changed_by': request.user})
-            if request.POST['value'] == 'disable':
-                user_account.account_status = UserStanding.ACCOUNT_DISABLED
-                messages.success(request, _(u"Successfully disabled {}'s account").format(user.username))
-            elif request.POST['value'] == 'reenable':
-                user_account.account_status = UserStanding.ACCOUNT_ENABLED
-                messages.success(request, _(u"Successfully reenabled {}'s account").format(username))
-            user_account.changed_by = request.user
-            user_account.standing_last_changed_at = timezone.now()
-            user_account.save()
-
-        elif request.POST['action'] == 'change-password':
-            user.set_password(request.POST['new-password'])
-            user.save()
-            messages.success(request, _(u"User password changed"))
-
-        return redirect('backoffice:user-list')
 
+    if 'action' in request.POST:
+        try:
+            user_actions[request.POST['action']](request, user)
+        except KeyError:
+            messages.error(request, _(u"Invalid user action."))
```

Comment:
Si user_actions lève une KeyError, on ne pourra pas la voir. Là aussi je suggère de réduire le scope du `try`.

    user_action = user_actions.get(request.POST.get('action'))
    if not user_action:
        messages.error(request, _(u"Invalid user action."))
    else:
        user_action(request, user)